File: apps/corporate-website/api/backend/app/routes/design_test.py
"""Routes for the Design Sense assessment API."""

from __future__ import annotations

from datetime import datetime, timezone
from statistics import mean
from typing import Any

from fastapi import APIRouter, HTTPException, status

from ..db import (
    get_applicants_collection,
    get_attempts_collection,
    get_images_collection,
    get_questions_collection,
    get_supplemental_questions_collection,
)
from ..models import (
    ApplicantInput,
    ApplicantSummary,
    ImageDocument,
    ImageResult,
    PublicBoostQuestion,
    PublicImage,
    PublicQuestion,
    QuestionDocument,
    QuestionResult,
    QuestionType,
    BoostQuestionDocument,
    PublicSelectionQuestion,
    SelectionQuestionDocument,
    SubmissionBand,
    SubmissionMetadata,
    SubmissionResult,
    TestSubmissionRequest,
)

router = APIRouter(prefix="/api/design-test", tags=["design-test"])


def _deserialize_question(document: dict[str, Any]) -> QuestionDocument:
    """Convert a raw MongoDB document into a strongly typed QuestionDocument."""
    payload = dict(document)
    object_id = payload.pop("_id", None)
    if object_id is not None:
        payload["id"] = str(object_id)
    return QuestionDocument.model_validate(payload)


def _question_to_public(question: QuestionDocument) -> PublicQuestion:
    """Strip private fields to expose a public-friendly question payload."""
    public_images = [
        PublicImage(
            imageId=image.imageId,
            src=image.src,
            imageType=image.imageType,
            displayLabel=image.displayLabel,
        )
        for image in question.images
    ]
    return PublicQuestion(
        questionNumber=question.questionNumber,
        questionType=question.questionType,
        images=public_images,
    )


async def _fetch_supplemental_docs() -> dict[int, SelectionQuestionDocument | BoostQuestionDocument]:
    """Load supplemental selection/boost questions keyed by question number."""
    collection = await get_supplemental_questions_collection()
    supplemental: dict[int, SelectionQuestionDocument | BoostQuestionDocument] = {}
    async for raw in collection.find():
        kind = raw.get("kind")
        if kind == "selection":
            doc = SelectionQuestionDocument.model_validate(raw)
        elif kind == "boost":
            doc = BoostQuestionDocument.model_validate(raw)
        else:
            continue
        supplemental[doc.questionNumber] = doc
    return supplemental


def _selection_to_public(doc: SelectionQuestionDocument) -> PublicSelectionQuestion:
    """Expose a selection question in the public API schema."""
    return PublicSelectionQuestion(
        questionNumber=doc.questionNumber,
        kind="selection",
        prompt=doc.prompt,
        options=doc.options,
    )


def _boost_to_public(doc: BoostQuestionDocument) -> PublicBoostQuestion:
    """Expose a boost question in the public API schema."""
    return PublicBoostQuestion(
        questionNumber=doc.questionNumber,
        kind="boost",
        prompt=doc.prompt,
        options=doc.options,
    )


@router.get("/questions", response_model=list[PublicQuestion])
async def list_questions() -> list[PublicQuestion]:
    """Return all assessment questions in display order."""
    collection = await get_questions_collection()
    cursor = collection.find().sort("questionNumber", 1)
    questions: list[PublicQuestion] = []
    async for doc in cursor:
        question = _deserialize_question(doc)
        questions.append(_question_to_public(question))
    return questions


@router.get("/supplemental", response_model=list[PublicSelectionQuestion | PublicBoostQuestion])
async def list_supplemental_questions() -> list[PublicSelectionQuestion | PublicBoostQuestion]:
    """Return supplemental questions (selection + boost) in order."""
    docs = await _fetch_supplemental_docs()
    result: list[PublicSelectionQuestion | PublicBoostQuestion] = []
    for question_number in sorted(docs.keys()):
        doc = docs[question_number]
        if isinstance(doc, SelectionQuestionDocument):
            result.append(_selection_to_public(doc))
        elif isinstance(doc, BoostQuestionDocument):
            result.append(_boost_to_public(doc))
    return result


async def _ensure_applicant(applicant: ApplicantInput, *, now: datetime) -> tuple[dict[str, Any] | None, int]:
    """Fetch an existing applicant by email and return their next attempt number."""
    applicants = await get_applicants_collection()
    normalized_email = applicant.email.lower()
    existing = await applicants.find_one({"email": normalized_email})
    if existing:
        attempt_number = int(existing.get("attemptCount", 0)) + 1
        return existing, attempt_number
    return None, 1


async def _update_applicant(
    applicant: ApplicantInput,
    attempt_number: int,
    *,
    now: datetime,
) -> None:
    """Upsert applicant metadata with the latest attempt information."""
    applicants = await get_applicants_collection()
    normalized_email = applicant.email.lower()
    update_payload = {
        "name": applicant.name,
        "updatedAt": now,
        "attemptCount": attempt_number,
    }
    result = await applicants.update_one(
        {"email": normalized_email},
        {
            "$set": update_payload,
            "$setOnInsert": {
                "email": normalized_email,
                "createdAt": now,
            },
        },
        upsert=True,
    )
    if result.matched_count == 0 and result.upserted_id is None:
        raise RuntimeError("Failed to upsert applicant document")


def _compute_band(score: float) -> SubmissionBand:
    """Translate a score into a qualitative band."""
    return SubmissionBand.from_score(score)


def _mean(values: list[float]) -> float:
    """Safe mean helper that tolerates empty lists."""
    return float(mean(values)) if values else 0.0


@router.post("/submit", response_model=SubmissionResult, status_code=status.HTTP_201_CREATED)
async def submit_results(payload: TestSubmissionRequest) -> SubmissionResult:
    """Compute the applicant's score, persist the attempt, and return the summary."""
    now = datetime.now(timezone.utc)
    normalized_email = payload.applicant.email.lower()
    attempts_collection = await get_attempts_collection()

    existing_attempt = await attempts_collection.find_one(
        {"sessionId": payload.sessionId, "applicantEmail": normalized_email}
    )
    if existing_attempt:
        return SubmissionResult.model_validate(
            {
                "applicant": {
                    "name": existing_attempt.get("applicantName"),
                    "email": existing_attempt.get("applicantEmail"),
                },
                "attemptNumber": existing_attempt.get("attemptNumber", 1),
                "sessionId": existing_attempt.get("sessionId"),
                "overallCloseness": existing_attempt.get("overallCloseness", 0.0),
                "overallClosenessPct": existing_attempt.get("overallClosenessPct", 0.0),
                "band": existing_attempt.get("band", SubmissionBand.NEEDS_WORK.value),
            }
        )

    applicant_doc, attempt_number = await _ensure_applicant(payload.applicant, now=now)

    response_map = {item.imageId: item.selectedScore for item in payload.responses}
    choice_map = {item.questionNumber: item.optionId for item in payload.choices}

    if 9 not in choice_map or 10 not in choice_map:
        raise HTTPException(status_code=400, detail="Additional questions 9 and 10 are required.")

    supplemental_docs = await _fetch_supplemental_docs()

    questions_collection = await get_questions_collection()
    question_cursor = questions_collection.find().sort("questionNumber", 1)
    questions: list[QuestionDocument] = []
    async for document in question_cursor:
        questions.append(_deserialize_question(document))

    image_ids = list(response_map.keys())
    images_collection = await get_images_collection()
    images_cursor = images_collection.find({"imageId": {"$in": image_ids}})
    actual_score_map: dict[str, ImageDocument] = {}
    async for image_doc in images_cursor:
        actual_image = ImageDocument.model_validate(image_doc)
        actual_score_map[actual_image.imageId] = actual_image

    question_results: list[QuestionResult] = []
    all_closeness: list[float] = []
    all_errors: list[float] = []
    scenario_summary: dict[str, Any] | None = None

    for question in questions:
        images: list[ImageResult] = []
        question_closeness: list[float] = []
        question_errors: list[float] = []

        for image in question.images:
            response = response_map.get(image.imageId)
            actual = actual_score_map.get(image.imageId)

            if response is None or actual is None:
                images.append(
                    ImageResult(
                        imageId=image.imageId,
                        selectedScore=response,
                        actualScore=actual.actualScore if actual else None,
                        excluded=True,
                    )
                )
                continue

            error = abs(response - actual.actualScore)
            closeness = 1 - (error / 2)
            images.append(
                ImageResult(
                    imageId=image.imageId,
                    selectedScore=response,
                    actualScore=actual.actualScore,
                    error=error,
                    closeness=closeness,
                    excluded=False,
                )
            )
            question_closeness.append(closeness)
            question_errors.append(error)
            all_closeness.append(closeness)
            all_errors.append(error)

        question_results.append(
            QuestionResult(
                questionNumber=question.questionNumber,
                questionType=question.questionType,
                closeness=_mean(question_closeness) if question_closeness else None,
                mae=_mean(question_errors) if question_errors else None,
                images=images,
            )
        )

    selection_doc_raw = supplemental_docs.get(9)
    if not isinstance(selection_doc_raw, SelectionQuestionDocument):
        raise HTTPException(status_code=500, detail="Selection question metadata missing.")

    selection_option_map = {option.optionId: option for option in selection_doc_raw.options}
    scenario_option_id = choice_map[9]
    selected_option = selection_option_map.get(scenario_option_id)
    correct_option = selection_option_map.get(selection_doc_raw.correctOptionId)
    if selected_option is None or correct_option is None:
        raise HTTPException(status_code=400, detail="Invalid selection for question 9.")

    selected_value = selected_option.value
    correct_value = correct_option.value
    scenario_error = abs(selected_value - correct_value)
    scenario_closeness = 1 - (scenario_error / 2)
    all_errors.append(scenario_error)
    all_closeness.append(scenario_closeness)
    scenario_summary = {
        "questionNumber": 9,
        "selectedOption": scenario_option_id,
        "selectedLabel": selected_option.label,
        "selectedValue": selected_value,
        "correctValue": correct_value,
        "error": scenario_error,
        "closeness": scenario_closeness,
    }

    base_closeness = _mean(all_closeness)
    mae = _mean(all_errors)

    boost_doc_raw = supplemental_docs.get(10)
    if not isinstance(boost_doc_raw, BoostQuestionDocument):
        raise HTTPException(status_code=500, detail="Boost question metadata missing.")

    boost_option_map = {option.optionId: option for option in boost_doc_raw.options}
    boost_option_id = choice_map[10]
    boost_option = boost_option_map.get(boost_option_id)
    if boost_option is None:
        raise HTTPException(status_code=400, detail="Invalid selection for question 10.")

    boost_multiplier = 1 + boost_option.boost
    boosted_closeness = min(base_closeness * boost_multiplier, 1.0)
    overall_closeness = boosted_closeness
    band = _compute_band(overall_closeness)

    metadata_dict = payload.metadata.model_dump() if payload.metadata else None
    overall_closeness_pct = round(overall_closeness * 100, 1)

    result = SubmissionResult(
        applicant=ApplicantSummary(name=payload.applicant.name, email=payload.applicant.email),
        attemptNumber=attempt_number,
        sessionId=payload.sessionId,
        overallCloseness=overall_closeness,
        overallClosenessPct=overall_closeness_pct,
        band=band,
    )

    role_summary = {
        "questionNumber": 10,
        "selectedOption": boost_option_id,
        "selectedLabel": boost_option.label,
        "boost": boost_option.boost,
    }

    await attempts_collection.insert_one(
        {
            "applicantEmail": normalized_email,
            "applicantName": payload.applicant.name,
            "attemptNumber": attempt_number,
            "sessionId": payload.sessionId,
            "submittedAt": now,
            "overallCloseness": result.overallCloseness,
            "overallClosenessPct": result.overallClosenessPct,
            "baseCloseness": base_closeness,
            "mae": mae,
            "band": result.band.value,
            "imageQuestions": [
                {
                    "questionNumber": question.questionNumber,
                    "questionType": question.questionType.value,
                    "closeness": question.closeness,
                    "mae": question.mae,
                    "images": [
                        {
                            "imageId": image.imageId,
                            "selectedScore": image.selectedScore,
                            "actualScore": image.actualScore,
                            "error": image.error,
                            "closeness": image.closeness,
                            "excluded": image.excluded,
                        }
                        for image in question.images
                    ],
                }
                for question in question_results
            ],
            "scenarioQuestion": scenario_summary,
            "rolePreference": role_summary,
            "boostMultiplier": boost_multiplier,
            "metadata": metadata_dict,
            "submittedAtIso": now.isoformat(),
        }
    )

    await _update_applicant(payload.applicant, attempt_number, now=now)

    return result

File: apps/corporate-website/api/server/src/routes/designTest.ts
import { Router } from "express";
import { z } from "zod";

import { getCollection } from "../db/mongo";

type QuestionType = "homestyle" | "product";

type ImageDocument = {
  imageId: string;
  src: string;
  actualScore: number;
  imageType: QuestionType;
  displayLabel?: string | null;
  filename?: string | null;
  questionNumber?: number | null;
};

type QuestionDocument = {
  questionNumber: number;
  questionType: QuestionType;
  images: ImageDocument[];
};

type SelectionOption = {
  optionId: string;
  label: string;
  value: number;
};

type BoostOption = {
  optionId: string;
  label: string;
  boost: number;
};

type SelectionQuestionDocument = {
  questionNumber: number;
  kind: "selection";
  prompt: string;
  correctOptionId: string;
  options: SelectionOption[];
};

type BoostQuestionDocument = {
  questionNumber: number;
  kind: "boost";
  prompt: string;
  options: BoostOption[];
};

type SubmissionBand = "Excellent" | "Good" | "Needs Work";

type SubmissionResult = {
  applicant: {
    name: string;
    email: string;
  };
  attemptNumber: number;
  sessionId: string;
  overallCloseness: number;
  overallClosenessPct: number;
  band: SubmissionBand;
};

const ApplicantInputSchema = z
  .object({
    name: z.string().min(2),
    email: z.string().email(),
    age: z.number().int().min(13).max(120).optional(),
    role: z.string().max(120).optional(),
    portfolioUrl: z.string().max(240).optional()
  })
  .strict();

const ResponseItemSchema = z
  .object({
    imageId: z.string().min(1),
    selectedScore: z.number().int().min(0).max(2)
  })
  .strict();

const ChoiceResponseSchema = z
  .object({
    questionNumber: z.number().int().min(1),
    optionId: z.string().min(1)
  })
  .strict();

const SubmissionMetadataSchema = z
  .object({
    startedAt: z.string().optional(),
    submittedAt: z.string().optional(),
    durationMs: z.number().int().min(0).optional(),
    userAgent: z.string().max(400).optional()
  })
  .passthrough();

const TestSubmissionRequestSchema = z
  .object({
    sessionId: z.string().min(6),
    applicant: ApplicantInputSchema,
    responses: z.array(ResponseItemSchema).min(1),
    choices: z.array(ChoiceResponseSchema).default([]),
    metadata: SubmissionMetadataSchema.optional()
  })
  .strict();

const router = Router();

const mean = (values: number[]): number => {
  if (!values.length) {
    return 0;
  }
  const total = values.reduce((sum, value) => sum + value, 0);
  return total / values.length;
};

const computeBand = (score: number): SubmissionBand => {
  if (score >= 0.85) {
    return "Excellent";
  }
  if (score >= 0.7) {
    return "Good";
  }
  return "Needs Work";
};

router.get("/questions", async (_req, res) => {
  try {
    const questionsCollection = await getCollection<QuestionDocument>("design_test_questions");
    const questions = await questionsCollection
      .find({}, { projection: { _id: 0 } })
      .sort({ questionNumber: 1 })
      .toArray();

    const payload = questions.map((question) => ({
      questionNumber: question.questionNumber,
      questionType: question.questionType,
      images: question.images.map((image) => ({
        imageId: image.imageId,
        src: image.src,
        imageType: image.imageType,
        displayLabel: image.displayLabel ?? null
      }))
    }));

    res.json(payload);
  } catch (error) {
    res.status(500).json({ detail: "Failed to load questions." });
  }
});

router.get("/supplemental", async (_req, res) => {
  try {
    const supplementalCollection = await getCollection<
      SelectionQuestionDocument | BoostQuestionDocument
    >("supplemental_questions");
    const docs = await supplementalCollection
      .find({}, { projection: { _id: 0 } })
      .sort({ questionNumber: 1 })
      .toArray();

    const payload = docs.map((doc) => {
      if (doc.kind === "selection") {
        return {
          questionNumber: doc.questionNumber,
          kind: "selection",
          prompt: doc.prompt,
          options: doc.options
        };
      }
      return {
        questionNumber: doc.questionNumber,
        kind: "boost",
        prompt: doc.prompt,
        options: doc.options
      };
    });

    res.json(payload);
  } catch (error) {
    res.status(500).json({ detail: "Failed to load supplemental questions." });
  }
});

router.post("/submit", async (req, res) => {
  const parsed = TestSubmissionRequestSchema.safeParse(req.body);
  if (!parsed.success) {
    res.status(400).json({ detail: "Invalid submission payload.", issues: parsed.error.issues });
    return;
  }

  const payload = parsed.data;
  const now = new Date();
  const normalizedEmail = payload.applicant.email.toLowerCase();

  try {
    const attemptsCollection = await getCollection<Record<string, unknown>>("attempts");
    const existingAttempt = await attemptsCollection.findOne({
      sessionId: payload.sessionId,
      applicantEmail: normalizedEmail
    });

    if (existingAttempt) {
      const result: SubmissionResult = {
        applicant: {
          name: String(existingAttempt.applicantName ?? payload.applicant.name),
          email: String(existingAttempt.applicantEmail ?? normalizedEmail)
        },
        attemptNumber: Number(existingAttempt.attemptNumber ?? 1),
        sessionId: String(existingAttempt.sessionId ?? payload.sessionId),
        overallCloseness: Number(existingAttempt.overallCloseness ?? 0),
        overallClosenessPct: Number(existingAttempt.overallClosenessPct ?? 0),
        band: (existingAttempt.band as SubmissionBand) ?? "Needs Work"
      };
      res.json(result);
      return;
    }

    const applicantsCollection = await getCollection<Record<string, unknown>>("applicants");
    const existingApplicant = await applicantsCollection.findOne({ email: normalizedEmail });
    const attemptNumber = existingApplicant
      ? Number(existingApplicant.attemptCount ?? 0) + 1
      : 1;

    const responseMap = new Map(payload.responses.map((item) => [item.imageId, item.selectedScore]));
    const choiceMap = new Map(payload.choices.map((item) => [item.questionNumber, item.optionId]));

    if (!choiceMap.has(9) || !choiceMap.has(10)) {
      res.status(400).json({ detail: "Additional questions 9 and 10 are required." });
      return;
    }

    const supplementalCollection = await getCollection<
      SelectionQuestionDocument | BoostQuestionDocument
    >("supplemental_questions");
    const supplementalDocs = await supplementalCollection.find({}, { projection: { _id: 0 } }).toArray();
    const supplementalMap = new Map<number, SelectionQuestionDocument | BoostQuestionDocument>();
    for (const doc of supplementalDocs) {
      supplementalMap.set(doc.questionNumber, doc);
    }

    const questionsCollection = await getCollection<QuestionDocument>("design_test_questions");
    const questions = await questionsCollection
      .find({}, { projection: { _id: 0 } })
      .sort({ questionNumber: 1 })
      .toArray();

    const imageIds = Array.from(responseMap.keys());
    const imagesCollection = await getCollection<ImageDocument>("images");
    const images = await imagesCollection
      .find({ imageId: { $in: imageIds } }, { projection: { _id: 0 } })
      .toArray();
    const actualScoreMap = new Map(images.map((image) => [image.imageId, image]));

    const questionResults: Array<Record<string, unknown>> = [];
    const allCloseness: number[] = [];
    const allErrors: number[] = [];

    for (const question of questions) {
      const imagesResult: Array<Record<string, unknown>> = [];
      const questionCloseness: number[] = [];
      const questionErrors: number[] = [];

      for (const image of question.images) {
        const response = responseMap.get(image.imageId);
        const actual = actualScoreMap.get(image.imageId);

        if (response === undefined || !actual) {
          imagesResult.push({
            imageId: image.imageId,
            selectedScore: response ?? null,
            actualScore: actual?.actualScore ?? null,
            excluded: true
          });
          continue;
        }

        const error = Math.abs(response - actual.actualScore);
        const closeness = 1 - error / 2;
        imagesResult.push({
          imageId: image.imageId,
          selectedScore: response,
          actualScore: actual.actualScore,
          error,
          closeness,
          excluded: false
        });

        questionCloseness.push(closeness);
        questionErrors.push(error);
        allCloseness.push(closeness);
        allErrors.push(error);
      }

      questionResults.push({
        questionNumber: question.questionNumber,
        questionType: question.questionType,
        closeness: questionCloseness.length ? mean(questionCloseness) : null,
        mae: questionErrors.length ? mean(questionErrors) : null,
        images: imagesResult
      });
    }

    const selectionDoc = supplementalMap.get(9);
    if (!selectionDoc || selectionDoc.kind !== "selection") {
      res.status(500).json({ detail: "Selection question metadata missing." });
      return;
    }

    const selectionOption = selectionDoc.options.find(
      (option) => option.optionId === choiceMap.get(9)
    );
    const correctSelection = selectionDoc.options.find(
      (option) => option.optionId === selectionDoc.correctOptionId
    );

    if (!selectionOption || !correctSelection) {
      res.status(400).json({ detail: "Invalid selection for question 9." });
      return;
    }

    const scenarioError = Math.abs(selectionOption.value - correctSelection.value);
    const scenarioCloseness = 1 - scenarioError / 2;
    allErrors.push(scenarioError);
    allCloseness.push(scenarioCloseness);

    const scenarioSummary = {
      questionNumber: 9,
      selectedOption: selectionOption.optionId,
      selectedLabel: selectionOption.label,
      selectedValue: selectionOption.value,
      correctValue: correctSelection.value,
      error: scenarioError,
      closeness: scenarioCloseness
    };

    const baseCloseness = mean(allCloseness);
    const mae = mean(allErrors);

    const boostDoc = supplementalMap.get(10);
    if (!boostDoc || boostDoc.kind !== "boost") {
      res.status(500).json({ detail: "Boost question metadata missing." });
      return;
    }

    const boostOption = boostDoc.options.find((option) => option.optionId === choiceMap.get(10));
    if (!boostOption) {
      res.status(400).json({ detail: "Invalid selection for question 10." });
      return;
    }

    const boostMultiplier = 1 + boostOption.boost;
    const overallCloseness = Math.min(baseCloseness * boostMultiplier, 1.0);
    const overallClosenessPct = Math.round(overallCloseness * 1000) / 10;
    const band = computeBand(overallCloseness);

    const result: SubmissionResult = {
      applicant: {
        name: payload.applicant.name,
        email: payload.applicant.email
      },
      attemptNumber,
      sessionId: payload.sessionId,
      overallCloseness,
      overallClosenessPct,
      band
    };

    const roleSummary = {
      questionNumber: 10,
      selectedOption: boostOption.optionId,
      selectedLabel: boostOption.label,
      boost: boostOption.boost
    };

    await attemptsCollection.insertOne({
      applicantEmail: normalizedEmail,
      applicantName: payload.applicant.name,
      attemptNumber,
      sessionId: payload.sessionId,
      submittedAt: now,
      overallCloseness: result.overallCloseness,
      overallClosenessPct: result.overallClosenessPct,
      baseCloseness,
      mae,
      band: result.band,
      imageQuestions: questionResults,
      scenarioQuestion: scenarioSummary,
      rolePreference: roleSummary,
      boostMultiplier,
      metadata: payload.metadata ?? null,
      submittedAtIso: now.toISOString()
    });

    const updateResult = await applicantsCollection.updateOne(
      { email: normalizedEmail },
      {
        $set: {
          name: payload.applicant.name,
          updatedAt: now,
          attemptCount: attemptNumber
        },
        $setOnInsert: {
          email: normalizedEmail,
          createdAt: now
        }
      },
      { upsert: true }
    );

    if (updateResult.matchedCount === 0 && !updateResult.upsertedId) {
      res.status(500).json({ detail: "Failed to upsert applicant document." });
      return;
    }

    res.status(201).json(result);
  } catch (error) {
    res.status(500).json({ detail: "Failed to submit results." });
  }
});

export default router;
